### ISSUE TITLE
Computed parameters for CustomNonbondedForce

### ISSUE THREAD

**Issue by peastman** [openmm/openmm#3368]
I've been thinking about ways to optimize #3361, and I thought of a feature that could be useful in many cases.  In that issue, a large part of the energy expression is really computing per-particle parameters.  In principle they only need to be computed once for each particle, but because of how CustomNonbondedForce works, they instead end up getting recomputed for each interaction.  That makes it slower than it needs to be.

What if we allowed you to specify "computed parameters"?  You would provide an expression that would get evaluated for each particle.  It could depend on global and per-particle parameters, but not on other computed parameters.  The computed parameter could then be used in the energy expression.

As an example, suppose you're doing an alchemical simulation.  You have a global parameter `lambda` that is used to interpolate between two per-particle parameters `sigma1` and `sigma2`.  You could write

```python
force.addComputedParameter('sigma', 'lambda*sigma1 + (1-lambda)*sigma2')
```

You could then use `sigma` in the energy expression.  This would remove the interpolation from the inner loop.  It also could save memory in the interaction kernel by only needing to load one parameter (`sigma`) in place of two (`sigma1` and `sigma2`).

**Comment by peastman** [comment]
Actually, I just realized this is basically the same as what CustomGBForce already does with `addComputedValue()`.  The only difference is that it wouldn't support computing parameters as sums over pairs.

**Comment by zhang-ivy** [comment]
This seems like it could be very helpful. 

>Actually, I just realized this is basically the same as what CustomGBForce already does with addComputedValue(). The only difference is that it wouldn't support computing parameters as sums over pairs.

But `CustomGBForce` is different enough from `CustomNonbondedForce` (i.e. doesn't have a `setLongRangeCorrection()` attribute), meaning it would be worth implementing this feature, right?

**Comment by peastman** [comment]
Correct.

**Comment by zhang-ivy** [comment]
@peastman : I didn't have a chance to ask you about this during our call today, but will you have time to implement this in the next week? I was hoping to have this feature available before I try to reduce the complexity of the custom expression, but if you don't expect it to done soon, I won't wait for it.

**Comment by peastman** [comment]
This is a major feature, so it will take a bit of time.  Also, I'm on vacation the next two weeks.

**Comment by jchodera** [comment]
Just a small addition: It would be helpful to be able to define derived global parameters too, since we will otherwise generally be computing a lot of derived global parameters over and over again in each pair computation.

**Comment by peastman** [comment]
Can you give an example?

**Comment by zhang-ivy** [comment]
@peastman : Happy new year! Just wondering when you are aiming to have this feature implemented?

**Comment by peastman** [comment]
I just started work on it earlier today.  I'll let you know how it goes.

**Comment by zhang-ivy** [comment]
Great!

**Comment by peastman** [comment]
I have the reference implementation working.  If you want to look at it or try it out, it's on the `computedvalues` branch at https://github.com/peastman/openmm.

**Comment by jchodera** [comment]
>>  It would be helpful to be able to define derived global parameters too, since we will otherwise generally be computing a lot of derived global parameters over and over again in each pair computation.
> Can you give an example?

I neglected to provide an example earlier, but here's a specific case where this would be useful:

We have a set of functions we define for how various global parameters defined for different forces are slaved to a single global parameter `lambda`:
```
    default_functions = {'lambda_rest_bonds': lambda x, beta0, beta : -2 * (1 - np.sqrt(beta / beta0)) * x + 1 if x < 0.5 else 2 * (1 - np.sqrt(beta / beta0)) * x - 1 + 2 * np.sqrt(beta / beta0),
                         'lambda_rest_angles': lambda x, beta0, beta : -2 * (1 - np.sqrt(beta / beta0)) * x + 1 if x < 0.5 else 2 * (1 - np.sqrt(beta / beta0)) * x - 1 + 2 * np.sqrt(beta / beta0),
                         'lambda_rest_torsions':lambda x, beta0, beta : -2 * (1 - np.sqrt(beta / beta0)) * x + 1 if x < 0.5 else 2 * (1 - np.sqrt(beta / beta0)) * x - 1 + 2 * np.sqrt(beta / beta0),
                         'lambda_rest_electrostatics': lambda x, beta0, beta : -2 * (1 - np.sqrt(beta / beta0)) * x + 1 if x < 0.5 else 2 * (1 - np.sqrt(beta / beta0)) * x - 1 + 2 * np.sqrt(beta / beta0),
                         'lambda_rest_electrostatics_exceptions': lambda x, beta0, beta : -2 * (1 - np.sqrt(beta / beta0)) * x + 1 if x < 0.5 else 2 * (1 - np.sqrt(beta / beta0)) * x - 1 + 2 * np.sqrt(beta / beta0),
                         'lambda_rest_sterics':lambda x, beta0, beta : -2 * (1 - np.sqrt(beta / beta0)) * x + 1 if x < 0.5 else 2 * (1 - np.sqrt(beta / beta0)) * x - 1 + 2 * np.sqrt(beta / beta0),
                         'lambda_rest_sterics_exceptions': lambda x, beta0, beta : -2 * (1 - np.sqrt(beta / beta0)) * x + 1 if x < 0.5 else 2 * (1 - np.sqrt(beta / beta0)) * x - 1 + 2 * np.sqrt(beta / beta0),
                         'lambda_alchemical_bonds_old': lambda x: 1 - x,
                         'lambda_alchemical_bonds_new': lambda x: x,
                         'lambda_alchemical_angles_old': lambda x: 1 - x,
                         'lambda_alchemical_angles_new': lambda x: x,
                         'lambda_alchemical_torsions_old': lambda x: 1 - x,
                         'lambda_alchemical_torsions_new': lambda x: x,
                         'lambda_alchemical_electrostatics_old': lambda x: 1 - x,
                         'lambda_alchemical_electrostatics_new': lambda x: x,
                         'lambda_alchemical_electrostatics_exceptions_old': lambda x: 1 - x,
                         'lambda_alchemical_electrostatics_exceptions_new': lambda x: x,
                         'lambda_alchemical_electrostatics_reciprocal': lambda x: x,
                         'lambda_alchemical_sterics_old': lambda x: 1 - x,
                         'lambda_alchemical_sterics_new': lambda x: x,
                         'lambda_alchemical_sterics_exceptions_old': lambda x: 1 - x,
                         'lambda_alchemical_sterics_exceptions_new': lambda x: x
                         }
```
Currently, our options for doing this are:
1. **Include all of these expressions in the `CustomIntegrator`** to compute `lambda` every time we need to update it (often multiple times for timestep); this is complex and can slow things down because no force computations can occur while the integrator is updating these global parameters.
2. **Fold the relevant expressions in every `CustomNonbondedForce interaction:** This is also slow, since it must be computed every interaction.
3. **Perform these computations in Python outside of the integration step:** This is not always practical unless we are doing some of the nonequilibrium integration in Python and some of it with an OpenMM integrator, which becomes complex

Being able to have CustomNonbondedForce compute global parameters from other global parameters avoids the need to include this computation in each pair computation.

It's possible that the implementation of `force.addComputedParameter` is sufficient, though, since it's just O(N) computations that can be performed in parallel---probably little difference from adding an explicit global parameter precomputation step!

**Comment by zhang-ivy** [comment]
>I have the reference implementation working.

Great! I just reviewed it and it looks good to me. I still haven't gotten around to figuring out how to compile OpenMM from source (though @jchodera recently got this working on our hpc, so he may have some pointers), so we may have to do the thing where I wait for you to merge it (so I can then use the nightly dev version to test).

**Comment by peastman** [comment]
All done.  The PR is #3412.  Can you take a look and try it out?

**Comment by zhang-ivy** [comment]
Wonderful! Could you merge this so I can install via the nightly dev build?

**Comment by peastman** [comment]
We really need to figure out how you can compile from source.  PRs are supposed to be reviewed before they're merged, not after.

**Comment by peastman** [comment]
This was implemented in #3412.
